tools: refactor update-authors.js to ESM

F3n67u opened the pull request:
<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

## Motivation

1. make `tools` directory's js file format more consistent
2. shifting towards ESM which is the official standard.

aduh95 commented on tools/update-authors.mjs:
```diff
@@ -33,7 +35,7 @@ output.write('# Authors ordered by first contribution.\n\n');
 const mailmap = new CaseIndifferentMap();
 {
-  const lines = fs.readFileSync(path.resolve(__dirname, '../', '.mailmap'),
+  const lines = fs.readFileSync(path.resolve(dirname, '../', '.mailmap'),
```
```suggestion
  const lines = fs.readFileSync(new URL('../'.mailmap', import.meta.url)),
```

aduh95 commented on tools/update-authors.mjs:
```diff
@@ -1,11 +1,13 @@
+import path from 'node:path';
+import fs from 'node:fs';
+import readline from 'node:readline';
+import { fileURLToPath } from 'url';
+
+const dirname = path.dirname(fileURLToPath(import.meta.url));
```
We probably don't need that.
```suggestion
```

aduh95 reviewed (commented):
Is there a reason to refactor this to ESM?

F3n67u commented:
> Is there a reason to refactor this ESM?

I see the newly added file on `tools` directory is all ESM module, so I think we could also refactor our commonjs format to esm module format to make:

1. all `tools` js file is more consistent: they are all esm module
2. shifting towards ESM which is the official standard.

aduh95 commented:
FYI ESM means 'ECMAScript module', so "esm module" is redundant. I'd suggest using `refactor … to ESM` or `refactor … to ESM syntax` instead.

Maybe instead of `rl.line`, we should use `for await(const line of rl)`, using to our advantage top-level await to improve the script readability would be a very good argument for switching to ESM.

F3n67u commented:
> FYI ESM means 'ECMAScript module', so "esm module" is redundant. I'd suggest using `refactor … to ESM` or `refactor … to ESM syntax` instead.
> 
> Maybe instead of `rl.line`, we should use `for await(const line of rl)`, using to our advantage top-level await to improve the script readability would be a very good argument for switching to ESM.

Good suggestion. I have refactored `rl.on('line'` to `for await(const line of rl)`. please take another look. the pr title is also changed to `refactor … to ESM`
